src/peer/socket.rs

```diff
@@ -99,103 +91,107 @@ async fn retainer<S: Signer>(
 
         u.set_query(Some(token.as_str()));
 
+        log::info!("connecting to relay {:?}", &u.domain());
         let (ws, _) = match tokio_tungstenite::connect_async(&u).await {
             Ok(v) => v,
             Err(err) => {
-                log::trace!(
+                log::error!(
                     "failed to establish connection to {:?} : {:#}",
                     u.domain(),
                     err
                 );
                 tokio::time::sleep(Duration::from_secs(2)).await;
-                log::trace!("retrying connection");
+                log::info!("retrying connection to {:?}", &u.domain());
                 continue;
             }
         };
         log::info!("now connected to relay {:?}", &u.domain());
 
-        let (mut write, read) = ws.split();
-        let mut read = read_stream(read);
-        let mut last = Instant::now();
+        let (mut write, mut read) = ws.split();
+        let ts = Arc::new(AtomicU64::new(timestamp()));
+
+        let down_tx = down_tx.clone();
+        let (close, mut closed) = mpsc::channel(1);
+
+        let ts_clone = Arc::clone(&ts);
+        let handler = tokio::spawn(async move {
+            let ts = ts_clone;
+            while let Some(message) = read.next().await {
+                let message = match message {
+                    Ok(message) => message,
+                    Err(err) => {
+                        if close.send(err).await.is_err() {
+                            log::error!("failed to notify of socket connection loss");
+                        }
+                        return;
```

Comment:
q: what suppose to happens after returning here

Reply:
this means the `read` routing is failing to read next message. we first notify the main loop that the read has failed (by sending the error on the close channel)

That send should never fail unless the retainer has exited completely, but in both cases the read routine need to exit.

The main loop will then re connect, start a new routing and wait for new messages